src/hashset.h

```diff
@@ -0,0 +1,188 @@
+    /* Allow a hashset to carry extra caller-defined metadata. The extra memory
+     * is initialized to 0. */
+    size_t (*getMetadataSize)(void);
+    /* Flag to disable incremental rehashing */
+    unsigned instant_rehashing : 1;
+} hashsetType;
+
+typedef enum {
+    HASHSET_RESIZE_ALLOW = 0,
+    HASHSET_RESIZE_AVOID,
+    HASHSET_RESIZE_FORBID,
+} hashsetResizePolicy;
+
+typedef void (*hashsetScanFunction)(void *privdata, void *element);
+
+/* Constants */
+#define HASHSET_BUCKET_SIZE 64 /* bytes, the most common cache line size */
+
+/* Scan flags */
+#define HASHSET_SCAN_EMIT_REF (1 << 0)
+
+/* Iterator. To be able to use a stack-allocated iterator, it's not opaque. The
+ * fields in this struct should not be accessed though. Just use the hashset
+ * iterator functions. */
+typedef struct {
+    hashset *hashset;
+    hashsetBucket *bucket;
+    long index;
+    int table;
+    int pos_in_bucket;
+    uint64_t safe : 1;
+    union {
+        /* Unsafe iterator fingerprint for misuse detection. */
+        uint64_t fingerprint : 63;
+        /* Safe iterator temporary storage for bucket chain compaction. */
+        uint64_t last_seen_size : 63;
+    };
+} hashsetIterator;
+
+/* Position, used by some hashset functions such as two-phase insert and delete.
+ * This is not opaque, to be able to place it on the stack, but don't access the
+ * fields. Only use it with the hashset functions. */
+typedef struct {
+    void *bucket;
+    uint16_t pos_in_bucket;
+    uint16_t table_index;
+} hashsetPosition;
+
+/* --- Prototypes --- */
+
+/* Hash function (global seed) */
+void hashsetSetHashFunctionSeed(const uint8_t *seed);
+uint8_t *hashsetGetHashFunctionSeed(void);
+uint64_t hashsetGenHashFunction(const char *buf, size_t len);
+uint64_t hashsetGenCaseHashFunction(const char *buf, size_t len);
+
+/* Global resize policy */
+void hashsetSetResizePolicy(hashsetResizePolicy policy);
+
+/* Hashset instance */
+hashset *hashsetCreate(hashsetType *type);
+void hashsetRelease(hashset *s);
+void hashsetEmpty(hashset *s, void(callback)(hashset *));
+hashsetType *hashsetGetType(hashset *s);
+void *hashsetMetadata(hashset *s);
+size_t hashsetSize(hashset *s);
+size_t hashsetBuckets(hashset *s);
+size_t hashsetChainedBuckets(hashset *s, int table);
+size_t hashsetMemUsage(hashset *s);
+void hashsetPauseAutoShrink(hashset *s);
+void hashsetResumeAutoShrink(hashset *s);
+int hashsetIsRehashing(hashset *s);
+int hashsetIsRehashingPaused(hashset *s);
+void hashsetRehashingInfo(hashset *s, size_t *from_size, size_t *to_size);
+int hashsetRehashMicroseconds(hashset *s, uint64_t us);
+int hashsetExpand(hashset *s, size_t size);
+int hashsetTryExpand(hashset *s, size_t size);
+int hashsetExpandIfNeeded(hashset *s);
+int hashsetShrinkIfNeeded(hashset *s);
+hashset *hashsetDefragTables(hashset *s, void *(*defragfn)(void *));
+
+/* Elements */
+int hashsetFind(hashset *s, const void *key, void **found);
+void **hashsetFindRef(hashset *s, const void *key);
```

Comment:
To see where it's used, see https://github.com/valkey-io/valkey/pull/1178, especially `db.c`.

It is to avoid looking up an element twice. The caller can replace an element using `*ref = new_element`. I agree it dangerous, also because the caller can set an element for the wrong key in the wrong place, but it's a simple API and it seems important to provide lookup + replace without another lookup.

In the dict, we often have the dictEntry and we replace the value in it. This way, the dict allows replace without a new lookup.

> I recommend changing this to return a `hashsetPosition` instead of a `void**`. Then you can have something like `hashsetGetAtPosition(position)` and `hashsetReplaceAtPosition(position, newElement)`. You can use epoch information in the `hashsetPosition` to guarantee safety, and also perform checks like `compare(ele1, ele2)` to ensure that the structure is still logically correct.

This sounds safer, but would it add any performance overhead?

This still doesn't prevent the caller from setting an element for the wrong, unless we hash the key and check it, so it still doesn't make it completely safe.

Reply:
I've used `hashsetFindRef` for defragmenting elements that are referenced by multiple data structures. (for example, skiplist/hashset in ZSET, or hashset/hashset in pubsub.) Defragmentation operations should be done atomically and with no multithreaded access. I guess hashsetPosition could be used without the additional checks in this case and others where we're certain our task is effectively atomic. 🤔

Reply:
Yes, exactly.  Using the hashsetPosition is essentially the same as using a pointer - but now you have control.  You haven't exposed your internals.  When the `hashsetPosition` is used, you can check a change ID on the hashset to ensure that no changes have occurred since the position was returned.  This will catch bad code - rather than silently ignoring it.

And yes, it's more expensive, but maybe during testing, we could have a check that fires the `compare` callback during a replace to ensure that the new value compares identically to the old one.  Another option would be to fire the function that computes the hash value - that could be essentially free if the application caches the hash value in the robj for large keys.

Reply:
@JimB123 If we provide a function `setElementAtPosition(hashset *s, hashsetPosition *p, void *element)`, this function can check that the position is still valid (that the epoch hasn't changed) which is good. But should this function also check that the element hashes to the same position?

If not, the user can mess up by setting an element in a mismatching position so it can't be looked up later. If we want verify this, then the function is almost as expensive as `hashsetReplace`, so it doesn't make much sense to provide `setElementAtPosition`. The user can just use `hashsetReplace` instead.

This is safety vs performance. I didn't want hashset to be any less performant than dict.

Reply:
My recommendation would be to perform a `compare` in development builds (for safety), but eliminate it in production builds, for performance.  This would just be an assertion wrapped in an ifdef.

Reply:
Follow-up: #1365